core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala

```diff
@@ -16,233 +16,1521 @@
+      val followerThrottles = calculateFollowerThrottles(moveMap)
+      modifyTopicThrottles(adminClient, leaderThrottles, followerThrottles)
+      val reassigningBrokers = calculateReassigningBrokers(moveMap)
+      val movingBrokers = calculateMovingBrokers(proposedReplicas.keySet.toSet)
+      modifyBrokerThrottles(adminClient,
+        reassigningBrokers, interBrokerThrottle,
+        movingBrokers, replicaAlterLogDirsThrottle)
+      if (interBrokerThrottle >= 0) {
+        println(s"The inter-broker throttle limit was set to ${interBrokerThrottle} B/s")
+      }
+      if (replicaAlterLogDirsThrottle >= 0) {
+        println(s"The replica-alter-dir throttle limit was set to ${replicaAlterLogDirsThrottle} B/s")
+      }
+    }
+
+    // Execute the partition reassignments.
+    val errors = alterPartitionReassignments(adminClient, proposedParts)
+    if (!errors.isEmpty) {
+      throw new TerseReassignmentFailureException(
+        "Error reassigning partition(s):%n%s".format(
+          errors.keySet.toBuffer.sortWith(compareTopicPartitions).map {
+            case part => s"${part}: ${errors(part).getMessage}"
+          }.mkString(System.lineSeparator())))
+    }
+    println(s"Started ${proposedParts.size} partition reassignment(s)")
 
-    // If there is an existing rebalance running, attempt to change its throttle
-    if (zkClient.reassignPartitionsInProgress()) {
-      println("There is an existing assignment running.")
-      reassignPartitionsCommand.maybeLimit(throttle)
+    if (!proposedReplicas.isEmpty) {
+      executeMoves(adminClient, proposedReplicas, timeoutMs, time)
+    }
+  }
+
+  /**
+   * Execute some partition log directory movements.
+   *
+   * @param adminClient                 The AdminClient to use.
+   * @param proposedReplicas            A map from TopicPartitionReplicas to the
+   *                                    directories to move them to.
+   * @param timeoutMs                   The maximum time in ms to wait for log directory
+   *                                    replica assignment to begin.
+   * @param time                        The Time object to use.
+   */
+  def executeMoves(adminClient: Admin,
+                   proposedReplicas: Map[TopicPartitionReplica, String],
+                   timeoutMs: Long,
+                   time: Time): Unit = {
+    val startTimeMs = time.milliseconds()
+    val pendingReplicas = new mutable.HashMap[TopicPartitionReplica, String]()
+    pendingReplicas ++= proposedReplicas
+    var done = false
+    do {
+      pendingReplicas --= alterReplicaLogDirs(adminClient, pendingReplicas)
+      if (pendingReplicas.isEmpty) {
+        println(s"Started ${proposedReplicas.size} replica log directory reassignment(s)")
+        done = true
+      } else if (time.milliseconds() >= startTimeMs + timeoutMs) {
+        throw new TerseReassignmentFailureException(
+          "Timed out before replica log dir(s) could be reassigned:%n%s".format(
```

Comment:
If we get a `ReplicaNotAvailableException` for one replica throughout the whole `timeoutMs` duration , we'd print out that all movements failed. Shall we print `pendingReplicas`?

Reply:
That's a good idea.  In general, we should just print out all the partitions or log directories that were affected for each operation.